File: java/logbrew-java/src/main/java/co/logbrew/sdk/ProductTimeline.java
package co.logbrew.sdk;

import java.net.URI;
import java.util.LinkedHashMap;
import java.util.Locale;
import java.util.Map;

/**
 * Builders for app-owned product and network timeline milestones.
 *
 * <p>The helpers create normal action attributes with privacy-safe primitive
 * metadata. They do not patch HTTP clients, capture payloads, capture headers,
 * or collect visual session replay.</p>
 */
public final class ProductTimeline {
    private ProductTimeline() {
    }

    /**
     * Starts a product action builder with the required action name.
     */
    public static ProductAction productAction(String name) {
        return new ProductAction(name);
    }

    /**
     * Starts a network milestone builder with the required route template.
     */
    public static NetworkMilestone networkMilestone(String routeTemplate) {
        return new NetworkMilestone(routeTemplate);
    }

    /**
     * Builder for a user- or product-owned action milestone.
     */
    public static final class ProductAction {
        private final String name;
        private String status = "success";
        private String routeTemplate;
        private String sessionId;
        private String traceId;
        private String screen;
        private String funnel;
        private String step;
        private Map<String, ?> metadata;

        private ProductAction(String name) {
            this.name = name;
        }

        /**
         * Sets the action status. Defaults to {@code success}.
         */
        public ProductAction status(String status) {
            this.status = status;
            return this;
        }

        /**
         * Sets a route template. Query strings and hash fragments are removed.
         */
        public ProductAction routeTemplate(String routeTemplate) {
            this.routeTemplate = routeTemplate;
            return this;
        }

        /**
         * Sets an app-owned session identifier.
         */
        public ProductAction sessionId(String sessionId) {
            this.sessionId = sessionId;
            return this;
        }

        /**
         * Sets a trace identifier that correlates the action with backend work.
         */
        public ProductAction traceId(String traceId) {
            this.traceId = traceId;
            return this;
        }

        /**
         * Sets the current app screen or view name.
         */
        public ProductAction screen(String screen) {
            this.screen = screen;
            return this;
        }

        /**
         * Sets the funnel name for product-analysis workflows.
         */
        public ProductAction funnel(String funnel) {
            this.funnel = funnel;
            return this;
        }

        /**
         * Sets the product step name inside a funnel.
         */
        public ProductAction step(String step) {
            this.step = step;
            return this;
        }

        /**
         * Sets optional primitive metadata. Nested objects are rejected.
         */
        public ProductAction metadata(Map<String, ?> metadata) {
            this.metadata = Validation.copyMetadata(metadata);
            return this;
        }

        /**
         * Builds normal LogBrew action attributes with timeline metadata.
         */
        public ActionAttributes toActionAttributes() {
            Validation.requireNonEmpty("product action name", name);
            Validation.requireAllowedValue("product action status", status, LogBrewClient.ACTION_STATUSES);
            Map<String, Object> timeline = timelineMetadata("product.action", metadata);
            putIfPresent(timeline, "routeTemplate", sanitizeRouteTemplate(routeTemplate));
            putIfPresent(timeline, "sessionId", stringOrNull(sessionId));
            putIfPresent(timeline, "traceId", stringOrNull(traceId));
            putIfPresent(timeline, "screen", stringOrNull(screen));
            putIfPresent(timeline, "funnel", stringOrNull(funnel));
            putIfPresent(timeline, "step", stringOrNull(step));

            ActionAttributes attributes = ActionAttributes.create(name, status).metadata(timeline);
            attributes.toMap();
            return attributes;
        }
    }

    /**
     * Builder for app-owned API milestones.
     */
    public static final class NetworkMilestone {
        private final String routeTemplate;
        private String name;
        private String method;
        private String status;
        private Integer statusCode;
        private Double durationMs;
        private String sessionId;
        private String traceId;
        private Map<String, ?> metadata;

        private NetworkMilestone(String routeTemplate) {
            this.routeTemplate = routeTemplate;
        }

        /**
         * Sets a custom action name. Defaults to {@code network.method route}.
         */
        public NetworkMilestone name(String name) {
            this.name = name;
            return this;
        }

        /**
         * Sets the HTTP method. Defaults to {@code GET}.
         */
        public NetworkMilestone method(String method) {
            this.method = method;
            return this;
        }

        /**
         * Sets the action status. Defaults from {@code statusCode} when omitted.
         */
        public NetworkMilestone status(String status) {
            this.status = status;
            return this;
        }

        /**
         * Sets the HTTP response status code.
         */
        public NetworkMilestone statusCode(int statusCode) {
            this.statusCode = Integer.valueOf(statusCode);
            return this;
        }

        /**
         * Sets the request duration in milliseconds.
         */
        public NetworkMilestone durationMs(double durationMs) {
            this.durationMs = Double.valueOf(durationMs);
            return this;
        }

        /**
         * Sets an app-owned session identifier.
         */
        public NetworkMilestone sessionId(String sessionId) {
            this.sessionId = sessionId;
            return this;
        }

        /**
         * Sets a trace identifier that correlates the milestone with backend work.
         */
        public NetworkMilestone traceId(String traceId) {
            this.traceId = traceId;
            return this;
        }

        /**
         * Sets optional primitive metadata. Nested objects are rejected.
         */
        public NetworkMilestone metadata(Map<String, ?> metadata) {
            this.metadata = Validation.copyMetadata(metadata);
            return this;
        }

        /**
         * Builds normal LogBrew action attributes with timeline metadata.
         */
        public ActionAttributes toActionAttributes() {
            String route = requireRouteTemplate(routeTemplate);
            String normalizedMethod = normalizeHttpMethod(method);
            Integer code = validateStatusCode(statusCode);
            String resolvedStatus = status == null ? statusFromStatusCode(code) : status;
            Validation.requireAllowedValue("network milestone status", resolvedStatus, LogBrewClient.ACTION_STATUSES);
            Validation.requireNonNegativeNumber("network milestone durationMs", durationMs);
            String resolvedName = name == null || name.trim().isEmpty()
                ? "network." + normalizedMethod.toLowerCase(Locale.ROOT) + " " + route
                : name;

            Map<String, Object> timeline = timelineMetadata("network.milestone", metadata);
            putIfPresent(timeline, "routeTemplate", route);
            putIfPresent(timeline, "method", normalizedMethod);
            putIfPresent(timeline, "statusCode", code);
            putIfPresent(timeline, "durationMs", durationMs);
            putIfPresent(timeline, "sessionId", stringOrNull(sessionId));
            putIfPresent(timeline, "traceId", stringOrNull(traceId));

            ActionAttributes attributes = ActionAttributes.create(resolvedName, resolvedStatus).metadata(timeline);
            attributes.toMap();
            return attributes;
        }
    }

    private static Map<String, Object> timelineMetadata(
        String source,
        Map<String, ?> localMetadata
    ) {
        Map<String, Object> merged = new LinkedHashMap<>();
        merged.put("source", source);
        Map<String, Object> copied = Validation.copyMetadata(localMetadata);
        if (copied != null) {
            merged.putAll(copied);
        }
        return merged;
    }

    private static void putIfPresent(Map<String, Object> metadata, String key, Object value) {
        if (value != null) {
            metadata.put(key, value);
        }
    }

    private static String requireRouteTemplate(String routeTemplate) {
        String sanitized = sanitizeRouteTemplate(routeTemplate);
        Validation.requireNonEmpty("network milestone routeTemplate", sanitized);
        return sanitized;
    }

    private static String sanitizeRouteTemplate(String routeTemplate) {
        if (routeTemplate == null) {
            return null;
        }
        String trimmed = routeTemplate.trim();
        if (trimmed.isEmpty()) {
            return "";
        }
        try {
            URI uri = URI.create(trimmed);
            if (uri.isAbsolute() || uri.getRawAuthority() != null) {
                String path = uri.getRawPath();
                return path == null || path.isEmpty() ? "/" : path;
            }
        } catch (IllegalArgumentException ignored) {
            // Fall back to simple query/hash stripping for route templates that are not valid URIs.
        }
        int query = trimmed.indexOf('?');
        int hash = trimmed.indexOf('#');
        int cutoff = firstPresentIndex(query, hash);
        if (cutoff >= 0) {
            trimmed = trimmed.substring(0, cutoff);
        }
        return trimmed.isEmpty() ? "/" : trimmed;
    }

    private static int firstPresentIndex(int left, int right) {
        if (left < 0) {
            return right;
        }
        if (right < 0) {
            return left;
        }
        return Math.min(left, right);
    }

    private static String normalizeHttpMethod(String method) {
        String value = method == null ? "GET" : method.trim();
        if (value.isEmpty()) {
            throw new SdkException("validation_error", "network milestone method must be a non-empty string");
        }
        String normalized = value.toUpperCase(Locale.ROOT);
        if (!isValidHttpMethod(normalized)) {
            throw new SdkException("validation_error", "network milestone method must be a valid HTTP method");
        }
        return normalized;
    }

    private static boolean isValidHttpMethod(String method) {
        for (int index = 0; index < method.length(); index++) {
            char character = method.charAt(index);
            if (index == 0 && (character < 'A' || character > 'Z')) {
                return false;
            }
            boolean allowed = (character >= 'A' && character <= 'Z')
                || (character >= '0' && character <= '9')
                || character == '_'
                || character == '-';
            if (!allowed) {
                return false;
            }
        }
        return true;
    }

    private static Integer validateStatusCode(Integer statusCode) {
        if (statusCode == null) {
            return null;
        }
        if (statusCode.intValue() < 100 || statusCode.intValue() > 599) {
            throw new SdkException("validation_error", "network milestone statusCode must be an integer from 100 to 599");
        }
        return statusCode;
    }

    private static String statusFromStatusCode(Integer statusCode) {
        if (statusCode != null && statusCode.intValue() >= 400) {
            return "failure";
        }
        return "success";
    }

    private static String stringOrNull(String value) {
        return value == null || value.trim().isEmpty() ? null : value;
    }
}

File: dotnet/logbrew-dotnet/src/LogBrew/ProductTimeline.cs
using System;
using System.Collections.Generic;

namespace LogBrew
{
    public static class ProductTimeline
    {
        public static ProductActionBuilder ProductAction(string name)
        {
            return new ProductActionBuilder(name);
        }

        public static NetworkMilestoneBuilder NetworkMilestone(string routeTemplate)
        {
            return new NetworkMilestoneBuilder(routeTemplate);
        }
    }

    public sealed class ProductActionBuilder
    {
        private readonly string name;
        private string status = "success";
        private string? routeTemplate;
        private string? sessionId;
        private string? traceId;
        private string? screen;
        private string? funnel;
        private string? step;
        private Dictionary<string, object?> metadata = new Dictionary<string, object?>();

        internal ProductActionBuilder(string name)
        {
            this.name = name;
        }

        public ProductActionBuilder WithStatus(string value)
        {
            status = value;
            return this;
        }

        public ProductActionBuilder WithRouteTemplate(string value)
        {
            routeTemplate = TimelineMetadata.SanitizeRouteTemplate("product routeTemplate", value);
            return this;
        }

        public ProductActionBuilder WithSessionId(string value)
        {
            sessionId = TimelineMetadata.RequirePrimitiveLabel("sessionId", value);
            return this;
        }

        public ProductActionBuilder WithTraceId(string value)
        {
            traceId = TimelineMetadata.RequirePrimitiveLabel("traceId", value);
            return this;
        }

        public ProductActionBuilder WithScreen(string value)
        {
            screen = TimelineMetadata.RequirePrimitiveLabel("screen", value);
            return this;
        }

        public ProductActionBuilder WithFunnel(string value)
        {
            funnel = TimelineMetadata.RequirePrimitiveLabel("funnel", value);
            return this;
        }

        public ProductActionBuilder WithStep(string value)
        {
            step = TimelineMetadata.RequirePrimitiveLabel("step", value);
            return this;
        }

        public ProductActionBuilder WithMetadata(IDictionary<string, object?> value)
        {
            metadata = TimelineMetadata.CopyMetadata(value);
            return this;
        }

        public ActionAttributes ToActionAttributes()
        {
            var actionMetadata = TimelineMetadata.CreateMetadata("product_timeline", metadata);
            TimelineMetadata.AddIfNotNull(actionMetadata, "routeTemplate", routeTemplate);
            TimelineMetadata.AddIfNotNull(actionMetadata, "sessionId", sessionId);
            TimelineMetadata.AddIfNotNull(actionMetadata, "traceId", traceId);
            TimelineMetadata.AddIfNotNull(actionMetadata, "screen", screen);
            TimelineMetadata.AddIfNotNull(actionMetadata, "funnel", funnel);
            TimelineMetadata.AddIfNotNull(actionMetadata, "step", step);
            return ActionAttributes.Create(name, status).WithMetadata(actionMetadata);
        }
    }

    public sealed class NetworkMilestoneBuilder
    {
        private readonly string routeTemplate;
        private string? name;
        private string method = "GET";
        private string? status;
        private int? statusCode;
        private double? durationMs;
        private string? sessionId;
        private string? traceId;
        private Dictionary<string, object?> metadata = new Dictionary<string, object?>();

        internal NetworkMilestoneBuilder(string routeTemplate)
        {
            this.routeTemplate = TimelineMetadata.SanitizeRouteTemplate("network milestone routeTemplate", routeTemplate);
        }

        public NetworkMilestoneBuilder WithName(string value)
        {
            name = value;
            return this;
        }

        public NetworkMilestoneBuilder WithMethod(string value)
        {
            method = TimelineMetadata.NormalizeHttpMethod(value);
            return this;
        }

        public NetworkMilestoneBuilder WithStatus(string value)
        {
            status = value;
            return this;
        }

        public NetworkMilestoneBuilder WithStatusCode(int value)
        {
            statusCode = value;
            return this;
        }

        public NetworkMilestoneBuilder WithDurationMs(double value)
        {
            durationMs = value;
            return this;
        }

        public NetworkMilestoneBuilder WithSessionId(string value)
        {
            sessionId = TimelineMetadata.RequirePrimitiveLabel("sessionId", value);
            return this;
        }

        public NetworkMilestoneBuilder WithTraceId(string value)
        {
            traceId = TimelineMetadata.RequirePrimitiveLabel("traceId", value);
            return this;
        }

        public NetworkMilestoneBuilder WithMetadata(IDictionary<string, object?> value)
        {
            metadata = TimelineMetadata.CopyMetadata(value);
            return this;
        }

        public ActionAttributes ToActionAttributes()
        {
            TimelineMetadata.ValidateStatusCode(statusCode);
            TimelineMetadata.ValidateDurationMs(durationMs);
            var normalizedStatus = status ?? (statusCode.HasValue && statusCode.Value >= 400 ? "failure" : "success");
            var actionName = name ?? ("network." + TimelineMetadata.ToLowerAscii(method) + " " + routeTemplate);
            var actionMetadata = TimelineMetadata.CreateMetadata("network_timeline", metadata);
            actionMetadata["routeTemplate"] = routeTemplate;
            actionMetadata["method"] = method;
            TimelineMetadata.AddIfNotNull(actionMetadata, "statusCode", statusCode);
            TimelineMetadata.AddIfNotNull(actionMetadata, "durationMs", durationMs);
            TimelineMetadata.AddIfNotNull(actionMetadata, "sessionId", sessionId);
            TimelineMetadata.AddIfNotNull(actionMetadata, "traceId", traceId);
            return ActionAttributes.Create(actionName, normalizedStatus).WithMetadata(actionMetadata);
        }
    }

    internal static class TimelineMetadata
    {
        internal static Dictionary<string, object?> CopyMetadata(IDictionary<string, object?>? source)
        {
            var copied = Validation.CopyMetadata(source);
            var metadata = new Dictionary<string, object?>();
            if (copied == null)
            {
                return metadata;
            }

            foreach (var item in copied.Values)
            {
                metadata[item.Key] = item.Value;
            }

            return metadata;
        }

        internal static Dictionary<string, object?> CreateMetadata(string source, IReadOnlyDictionary<string, object?> appMetadata)
        {
            var metadata = new Dictionary<string, object?> { ["source"] = source };
            foreach (var item in appMetadata)
            {
                if (!string.Equals(item.Key, "source", StringComparison.Ordinal))
                {
                    metadata[item.Key] = item.Value;
                }
            }

            return metadata;
        }

        internal static void AddIfNotNull(Dictionary<string, object?> metadata, string key, object? value)
        {
            if (value != null)
            {
                metadata[key] = value;
            }
        }

        internal static string RequirePrimitiveLabel(string label, string value)
        {
            Validation.RequireNonEmpty(label, value);
            return value.Trim();
        }

        internal static string SanitizeRouteTemplate(string label, string routeTemplate)
        {
            Validation.RequireNonEmpty(label, routeTemplate);
            var trimmed = routeTemplate.Trim();
            if (Uri.TryCreate(trimmed, UriKind.Absolute, out var uri) && !string.IsNullOrEmpty(uri.Host))
            {
                return string.IsNullOrEmpty(uri.AbsolutePath) ? "/" : uri.AbsolutePath;
            }

            var queryIndex = trimmed.IndexOf('?');
            var fragmentIndex = trimmed.IndexOf('#');
            var cutoff = FirstPresentIndex(queryIndex, fragmentIndex);
            if (cutoff >= 0)
            {
                trimmed = trimmed.Substring(0, cutoff).TrimEnd();
            }

            return trimmed.Length == 0 ? "/" : trimmed;
        }

        internal static string NormalizeHttpMethod(string method)
        {
            if (string.IsNullOrWhiteSpace(method))
            {
                throw InvalidMethod();
            }

            var normalized = method.Trim().ToUpperInvariant();
            foreach (var character in normalized)
            {
                if (!IsMethodCharacter(character))
                {
                    throw InvalidMethod();
                }
            }

            return normalized;
        }

        internal static void ValidateStatusCode(int? statusCode)
        {
            if (statusCode.HasValue && (statusCode.Value < 100 || statusCode.Value > 599))
            {
                throw new SdkException("validation_error", "network milestone statusCode must be between 100 and 599");
            }
        }

        internal static void ValidateDurationMs(double? durationMs)
        {
            if (!durationMs.HasValue)
            {
                return;
            }

            Validation.RequireFiniteNumber("network milestone durationMs", durationMs.Value);
            if (durationMs.Value < 0)
            {
                throw new SdkException("validation_error", "network milestone durationMs must be non-negative");
            }
        }

        internal static string ToLowerAscii(string value)
        {
            var characters = value.ToCharArray();
            for (var index = 0; index < characters.Length; index++)
            {
                if (characters[index] >= 'A' && characters[index] <= 'Z')
                {
                    characters[index] = (char)(characters[index] + ('a' - 'A'));
                }
            }

            return new string(characters);
        }

        private static int FirstPresentIndex(int first, int second)
        {
            if (first < 0)
            {
                return second;
            }

            if (second < 0)
            {
                return first;
            }

            return Math.Min(first, second);
        }

        private static bool IsMethodCharacter(char character)
        {
            return (character >= 'A' && character <= 'Z')
                || (character >= '0' && character <= '9')
                || character == '-'
                || character == '_';
        }

        private static SdkException InvalidMethod()
        {
            return new SdkException("validation_error", "network milestone method must be a valid HTTP method");
        }
    }
}
